### PR TITLE
add timer models

### LINKED ISSUE #1: Need timer models
A timer represents the in memory configuration of a recurring task. There will need to be an in memory model ("struct") for this.

### PR DESCRIPTION
**Pull request by nivista**
closes #1

### CODE REVIEW

File: timer/timer.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		ID             uuid.UUID
+		Domain         string
+		ExecutionCount uint32
+		Meta           Meta
+		Task           Task
+		Schedule       Schedule
+	}
+
+	// Task represents a timer task.
+	Task interface {
```
Comment (andrewjdawson2016) [inline timer/timer.go]: This is weird to have an exported interface with its only method be non-exported.

File: timer/timer.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		isTask()
```
Comment (andrewjdawson2016) [inline timer/timer.go]: This is weird - the interface is named task and it has one method isTask? Not clear.

File: timer/timer.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	// Schedule represents the "when" configuration of a timer.
+	Schedule interface {
```
Comment (andrewjdawson2016) [inline timer/timer.go]: Same comments here.

File: timer/timer.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	Meta struct {
+		CreationTime time.Time
+	}
+)
+
+// HTTP Task represents an HTTP request.
+type (
+	HTTP struct {
+		URL string
+		Method
```
Comment (andrewjdawson2016) [inline timer/timer.go]: Why is method embeded but others are not?

File: timer/timer.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+func (HTTP) isTask() {}
+
+// Method definitions
+const (
+	GET Method = iota
+	POST
+)
+
+// Schedule definitions
+type (
```
Comment (andrewjdawson2016) [inline timer/timer.go]: There is too much going on in this file. It contains metadata for a timer, schedule, different timer types, tasks and timers. These need to be broken apart and tested separately.

File: timer/timer.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	var sched string
+	switch t.Schedule.(type) {
+	case *Cron:
+		sched = "cron"
+	case *Interval:
+		sched = "interval"
+	default:
+		panic("Can't marshal schedule, unknown type.")
+	}
+	m["scheduleType"] = sched
```
Comment (andrewjdawson2016) [inline timer/timer.go]: This is weird to marshal something with an interface and then add the interface back in. You probably just want your types to be serializable (I would actually just start with proto for these as well).

**Review by andrewjdawson2016 — CHANGES_REQUESTED** [review]
I think this code needs a few iterations. 

Main things 
- This file is doing a lot it likely should be pulled apart 
- The serialization + interfaces look weird 
- The exported interfaces with non-exported types is concerning 
- It is concerning that Timer has all fields exported -  this means no where in the code can you assert any invariant about the state of timer.

File: timer/timer.go (hunk 1 above)
Comment (nivista) [reply timer/timer.go]: This makes it so others can type switch on it but not implement it.

File: timer/timer.go (hunk 4 above)
Comment (nivista) [reply timer/timer.go]: Ok i'll make it not embedded

File: timer/timer.go (hunk 5 above)
Comment (nivista) [reply timer/timer.go]: How do I test the types, they're just structs/interfaces? I can test marshalling/unmarshalling.

File: timer/timer.go (hunk 6 above)
Comment (nivista) [reply timer/timer.go]: I don't understand this. Timers have to be marshalled and they also semantically have interface fields. Are you saying I should just have protobufs as an intermediate type? I've done that and it works, but the conversion from proto type to regular type is just as nasty, but ya I can do it. Or are you saying timers shouldn't have interface fields, in which case I'd like to understand what that would look like.

**Comment by nivista** [comment]
There was too much wrong with this initial PR, and changes I made to fix it involved changing protos which should be outside the scope of this PR. It'll be easier just to make new PRs.

**Pull request closed without merging**